apsuite/commisslib/measure_tbt_data.py

```diff
@@ -0,0 +1,209 @@
+            phase_guesses M-array: phase guess for each BPM
+
+        Returns:
+            params (3, M)-array: fitted ampliude, tune and phase for each BPM.
+                                each column corresponds to a BPM, each row
+                                corresponds to amplitudes, tune and phase
+        """
+        ilist = _np.arange(matrix.shape[0])
+        params = _np.zeros((3, matrix.shape[-1]))
+        for bpm_idx, bpm_data in enumerate(matrix.T):
+            p0=[amp_guesses[bpm_idx],tune_guess,phase_guesses[bpm_idx]]
+            popt, *_ = _curve_fit(f=self.harmonic_tbt_model,
+                                  xdata=ilist,ydata=bpm_data,
+                                  p0=p0)
+            params[:, bpm_idx] = popt
+        return params
+
+    def calculate_betafunc_and_action(self, amplitudes, nominal_beta):
+        """Calculates beta function and betatron action as in Eq. (9)
+            of Ref. [1]"""
+        action = _np.sum(amplitudes**4)
+        action /= _np.sum(amplitudes**2 * nominal_beta)
+        beta = amplitudes**2/action
+        return beta, action
+
+class TbTDataParams(_AcqBPMsSignalsParams):
+    """."""
+
+    def __init__(self, hkick=None, vkick=None):
+        super().__init__()
+        self.nrpoints_before = 100
+        self.nrpoints_after = 2000
+        self.acq_rate = 'TbT'
+        self.signals2acq = 'XYS'
+        self._hkick = hkick
+        self._vkick = vkick
+
+
+    def __str__(self):
+        stg = super().__str__()
+        ftmp = '{0:26s} = {1:9.6f}  {2:s}\n'.format
+        stg += ftmp('hkick', self.hkick, '[urad]')
+        stg += ftmp('vkick', self.vkick, '[urad]')
+        return stg
+
+    @property
+    def hkick(self):
+        """."""
+        return self._hkick
+
+    @hkick.setter
+    def hkick(self, val):
+        self._hkick = val
+
+    @property
+    def vkick(self):
+        """."""
+        return self._vkick
+
+    @vkick.setter
+    def vkick(self, val):
+        self._vkick = val
+
+class MeasureTbTData(_AcqBPMsSignals):
+    """."""
+    def __init__(self, filename='', isonline=False):
+        super.__init__(params=TbTDataParams(), isonline=isonline,
+                        ispost_mortem=False)
+        self._fname = filename
+
+
+    def create_devices(self):
+        super().create_devices()
+        self.devices['pingh'] = PowerSupplyPU(
+            PowerSupplyPU.DEVICES.SI_INJ_DPKCKR)
+        self.devices['pinghv'] = PowerSupplyPU(PowerSupplyPU.DEVICES.SI_PING_V)
+
+
+    @property
+    def fname(self):
+        """."""
+        return self._fname
+
+    @fname.setter
+    def fname(self, val):
+        self._fname = val
+
+    def prepare_pingers(self):
+        """."""
+        pingh, pingv = self.devices['pingh'], self.devices['pingv']
+        params = self.params
+        hkick, vkick = params['hkick'], params['vkick']
+        pingh.strength = hkick / 1e3   # [urad]
+        pingv.strength = vkick / 1e3   # [urad]
```

Comment:
It is good to keep our measurement classes as transparent as possible: before the measurement, store the values of parameters that will be changed, change it for the measurement, and then restore the initial state.

Reply:
Is it possible for the measurement to be interrupted ? If yes, @vellosok75 did you implement restoration in this situation, apart from restoration at completion ?